### DISCUSSION TITLE
Proper, reusable way to ensure that user is authenticated on specific routes?

### DISCUSSION THREAD

**Discussion by simplenotezy** [nextauthjs/next-auth#2200] [Help] [answered]
What is a proper, secure way, to ensure that user is authenticated on specific routes?

Say I have /admin/ page, on which I require the user to be authenticated, before I execute the code in **getServerSideProps**, I want to ensure that the user is logged in. If the user is not logged in, I want to redirect to the login page, and after login, user will be redirected back.

I can't find much documentation on this. I can implement my own workaround, but I expect there to be some default way of doing this, in the library.

Any help is much appreciated

**Comment by balazsorban44** [comment, marked as answer]
https://next-auth.js.org/tutorials/securing-pages-and-api-routes

Personally I don't like to use getServerSideProps for only auth, so I came up with this:

https://next-auth.js.org/getting-started/client#custom-client-session-handling

(I think we should move the latter into the former page)

I hope it helps! 😊

**Reply by simplenotezy** [reply]
Thanks for your very prompt reply. @balazsorban44 , however, it doesn't really solve my issue of creating a unified way to handle "protected" routes. I would have to duplicate a lot of code. Ideally, I imagine I create a middleware hook somewhere, that I can simply use on getServerSideProps, and which will automatically redirect the user. The documentation (and links you've sent) doesn't cover this unfortunately :/

**Reply by balazsorban44** [reply]
You could create a HOC and wrap your `getServerSideProps`, if you want to go that way.

My suggestion with the client-side redirect - after setting it up - would only require to do

`YourPage.auth = true` per page.

I don't know if there is a way of requiring auth in code less than this per page. 😅

**Reply by simplenotezy** [reply]
Interesting. That sound's like a cool approach. Can you elaborate a bit more, perhaps with what you mean with wrapping getServerSideProps or providing a minimal code example?

**Reply by balazsorban44** [reply]
Writing this from memory/not tested:

```js
function withAuth(getServerSideProps) {
  return async function (ctx) {
    const session = await getSession({ ctx })
    if (!session) {
      return {
        redirect: {
          to: `/login?redirect-to=${ctx.req.url}`,
        },
      }
    }
    const response = await getServerSideProps(ctx)
    return {
      ...response,
      props: { ...response.props, session },
    }
  }
}

// Use it like this on your page.
export cosnt getServerSideProps = withAuth((ctx) => {
  // Your usual getServerSideProps
})
```


Still, I would go with a client-side redirect, so you could show a skeleton UI for example. But it is up to you.

**Reply by simplenotezy** [reply]
Ah that's cool @balazsorban44. I wasn't even aware chaining/hoc in this way was possible. Exactly what I was looking at. I'll think more about sticking with 100% client-side redirect.

Would that mean that you wouldn't fetch any data on serverSide either? Just always fetch data from client? I see some benefit's here, e.g. 1) Only worry about data fetching one place (client) 2) You could export your web-app as 100% static.

**Reply by balazsorban44** [reply]
getServerSideProps is no magic, just a regular JS function. 😊

I would only fetch data on server side, if it is static/can be fetched through `getStaticProps`. It depends on the type of page you want to show though. If it's an admin dashboard, there is probably no static content, and I would do all the fetching clientside, like a regular SPA. At work, we use this approach for articles that need auth from our users, but we render SOME parts of the article (with a paywall on top) to give a sneak peek of our content. We render the public parts statically, and fetch the protected content client-side with fetch calls.

My view on getServerSideProps might be very opinionated, but I see almost no reason to use it.

**Reply by simplenotezy** [reply]
Hey @balazsorban44 I makes good sense. Yes, this is for an admin dashboard, so I guess theres no need to fetch anything on server. I like the idea of fetching one place only (client) as well. Thanks for your valuable input.

**Reply by simplenotezy** [reply]
@balazsorban44 Question regarding your "client-only" suggestion, which I really like. Where does `pageProps.session` come from? 

```
export default function App({ Component, pageProps }) {
  return (
    <Provider session={pageProps.session}>
```

**Reply by simplenotezy** [reply]
Also, while playing around with your suggestion I experience a weird issue when switching pages. It keeps saying "Loading user...", without actually doing anything. If I refresh the page, it works fine. Not sure what's going wrong here 😊 

![image](https://user-images.githubusercontent.com/3786627/123066694-98059500-d410-11eb-9267-ef9aa60c9ce1.png)

**Reply by simplenotezy** [reply]
Update: If I console.log like so:

```
useEffect(() => {
    console.log('is user?', isUser, session, loading);
    if (loading) return; // Do nothing while loading
    if (!isUser) signIn(); // If not authenticated, force log in
  }, [isUser, loading]);
```

On page switch I get:
`is user? false undefined true`

On page refresh I get:
`is user? true {user: {…}, expires: "2021-07-23T09:10:05.755Z", accessToken: "XXXXX"} false`

So it seems like "loading" state never changes.
